transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+		fds = 0;
+		/* Writeable monitor */
+		poll_fds[fds].fd = write_fd[0];
+		poll_fds[fds].events = POLLIN;
+		fds++;
+		if(pfd > 0) {
+			/* Janus API */
+			poll_fds[fds].fd = pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		if(admin_pfd > 0) {
+			/* Admin API */
+			poll_fds[fds].fd = admin_pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		/* Iterate on available clients, to see if we need to POLLIN or POLLOUT too */
+		janus_mutex_lock(&clients_mutex);
+		GHashTableIter iter;
+		gpointer value;
+		g_hash_table_iter_init(&iter, clients_by_fd);
+		while(g_hash_table_iter_next(&iter, NULL, &value)) {
+			janus_pfunix_client *client = value;
+			if(client->fd > 0) {
+				poll_fds[fds].fd = client->fd;
+				poll_fds[fds].events = POLLIN;
+				if(g_async_queue_length(client->messages) > 0) {
+					poll_fds[fds].events = POLLOUT;
+				}
+				fds++;
+			}
+		}
+		janus_mutex_unlock(&clients_mutex);
+
+		/* Start polling */
+		int res = poll(poll_fds, fds, 10000);
+		if(res == 0)
+			continue;
+		if(res < 0) {
+			JANUS_LOG(LOG_ERR, "poll() failed\n");
+			break;
+		}
+		for(int i=0; i<fds; i++) {
+			if((poll_fds[i].revents & POLLOUT) == POLLOUT) {
+				/* Find the client from its file descriptor */
+				janus_mutex_lock(&clients_mutex);
+				janus_pfunix_client *client = g_hash_table_lookup(clients_by_fd, GINT_TO_POINTER(poll_fds[i].fd));
+				if(client != NULL && client->fd > 0) {
+					char *payload = NULL;
+					while((payload = g_async_queue_try_pop(client->messages)) != NULL) {
+						int res = write(client->fd, payload, strlen(payload));
+						/* FIXME Should we check if sent everything? */
+						JANUS_LOG(LOG_HUGE, "Written %d/%zu bytes on %d\n", res, strlen(payload), client->fd);
+						g_free(payload);
+					}
+				}
+				janus_mutex_unlock(&clients_mutex);
+			}
+			if((poll_fds[i].revents & POLLIN) == POLLIN) {
+				if(poll_fds[i].fd == write_fd[0]) {
+					/* Read and ignore: we use this to unlock the poll if there's data to write */
+					res = read(poll_fds[i].fd, buffer, 4096);
+				} else if(poll_fds[i].fd == pfd) {
+					/* Janus API: accept the new client */
+					struct sockaddr_un address;
+					socklen_t address_length = 0;
+					int cfd = accept(pfd, (struct sockaddr *) &address, &address_length);
+					if(cfd > 0) {
+						JANUS_LOG(LOG_VERB, "Got new Unix Sockets Janus API client: %d\n", cfd);
+						/* Allocate new client */
+						janus_pfunix_client *client = g_malloc0(sizeof(janus_pfunix_client));
+						client->fd = cfd;
+						client->admin = FALSE;	/* Janus API client */
+						client->messages = g_async_queue_new();
+						client->session_timeout = 0;
+						/* Take note of this new client */
+						janus_mutex_lock(&clients_mutex);
+						g_hash_table_insert(clients_by_fd, GINT_TO_POINTER(cfd), client);
+						g_hash_table_insert(clients, client, client);
+						janus_mutex_unlock(&clients_mutex);
+					}
+				} else if(poll_fds[i].fd == admin_pfd) {
```

Comment:
You can avoid some duplication here by executing the same code if the fd is admin_fd or pfd, and derive the admin boolen from that.